deployment/common/changeset/state/evm.go

```
@@ -82,16 +92,81 @@ func MaybeLoadMCMSWithTimelockState(env cldf.Environment, chainSelectors []uint6
 	return result, nil
 }
 
+// MergeAddressesFromBothSourcesEVMWithQualifier combines addresses from both DataStore and AddressBook making it backward compatible.
+// This version supports qualifiers for filtering DataStore addresses.
+// When a qualifier is specified, only DataStore addresses with that qualifier are returned (no AddressBook merge)
+// to ensure isolation between different deployments.
+func MergeAddressesFromBothSourcesEVMWithQualifier(env cldf.Environment, chainSelector uint64, qualifier string) (map[string]cldf.TypeAndVersion, error) {
+	// If a qualifier is specified, only use DataStore to ensure isolation between deployments
+	if qualifier != "" {
+		if env.DataStore != nil {
+			return LoadAddressesFromDataStore(env.DataStore, chainSelector, qualifier)
+		}
+		return nil, fmt.Errorf("DataStore not available but qualifier %s specified", qualifier)
+	}
+
+	// For backward compatibility without qualifier, merge both sources
+	// Start with addresses from AddressBook
+	addressBookAddresses := make(map[string]cldf.TypeAndVersion)
+	if addresses, err := env.ExistingAddresses.AddressesForChain(chainSelector); err == nil {
+		addressBookAddresses = addresses
+	} else if !errors.Is(err, cldf.ErrChainNotFound) {
```

I take we don't return the error `ErrChainNotFound` cause we are expecting the datastore to maybe have addresses for this chain?

Do we really want to swallow this error?

---

actually this is maintaining the current behaviour [here](https://github.com/smartcontractkit/chainlink/pull/19151/files#diff-999c8d3419b9c7a8ae26c5213602878c03037d548004c7eb1cfa9ce28267558dL796-L800) in which if the  chain is not found it returns an empty slice of addresses, in our case it will also try to fill it up with whats on the datastore